### DISCUSSION TITLE
moor isolate to work with windows

### DISCUSSION THREAD

**Discussion by matthewkooshad** [simolus3/drift#1292] [Q&A] [answered]
could i get help with moor isolate as from https://github.com/simolus3/moor/discussions/1257#discussioncomment-855666, with trying to get the moor isolate to work on Windows? 

i'm wondering where do i put in the open snippet for sqlite3.dll from https://moor.simonbinder.eu/docs/other-engines/vm/#supported-platforms -- i'm trying as below, but it's not working. i'm using https://pub.dev/packages/sqlite3_library_windows , but using the snippet from https://moor.simonbinder.eu/docs/other-engines/vm/#supported-platforms gives the same result.
```dart
Future<void> initDatabase() async {
  open.overrideFor(OperatingSystem.windows, openSQLiteOnWindows);
  final isolate = await _createMoorIsolate();
  final connection = await isolate.connect();
  db = Database.connect(connection);
}
```

if i use https://pub.dev/packages/sqlite3_library_windows in a project like this, it works fine on windows for me:
```dart
Future<void> main() async {
  // Override the sqlite3 load config when OS is Windows
  open.overrideFor(OperatingSystem.windows, openSQLiteOnWindows);

  // Folder where the database will be stored
  final dbFolder = await getApplicationSupportDirectory();

  // If the database file doesn't exist, create it.
  File dbFile = await File(dbFolder.path + "\\sqlite3_library_windows_example\\db").create(recursive: true);

  // Open the database file
  db = sqlite3.open(dbFile.path);

  // Create 'count' table if the table doesn't already exist
  db.execute('CREATE TABLE IF NOT EXISTS '
      'count (count_value INTEGER NOT NULL);');

  runApp(MyApp());
}
```

so, i'm really just stuck with getting the moor isolate to work with windows.

**Comment by simolus3** [comment, marked as answer]
You also need to call `open.overrideFor(OperatingSystem.windows, openSQLiteOnWindows);` on the background isolate. This is because Dart isolates don't share memory or global variables, so you need to re-initialize them on the background isolate if you're not using the defaults. The right place to put them is in `_startBackground`, before you're actually opening the database.

I'll change the documentation to include this information:

> As the name implies, Dart isolates don't share memory. This means that global variables
and values accessible in one isolate may not be visible in a background isolate.
>
>For instance, if you're using `open.overrideFor` from `package:sqlite3`, you need to do that
on the isolate where you're actually opening the database!
With a background isolate as shown here, the right place to call `open.overrideFor` is in the
`_startBackground` function, before you're using `MoorIsolate.inCurrent`.
Other global fields that you might be relying on when constructing the database (service
locators like `get_it` come to mind) may also need to be initialized seperately on the background
isolate.

**Reply by matthewkooshad** [reply]
that worked great to move the statement there. it works now, thanks!
